src/stdio_mgr/types.py

```diff
@@ -0,0 +1,179 @@
+
+# AbstractContextManager was introduced in Python 3.6
+# and may be used with typing.ContextManager.
+try:
+    from contextlib import AbstractContextManager
+except ImportError:  # pragma: no cover
+
+    class AbstractContextManager(abc.ABC):
+        """An abstract base class for context managers."""
+
+        def __enter__(self):
+            """Return `self` upon entering the runtime context."""
+            return self
+
+        @abc.abstractmethod
+        def __exit__(self, exc_type, exc_value, traceback):
+            """Raise any exception triggered within the runtime context."""
+            return None
+
+        @classmethod
+        def __subclasshook__(cls, subclass):
+            """Check whether subclass is considered a subclass of this ABC."""
+            if cls is AbstractContextManager:
+                return _collections_abc._check_methods(
+                    subclass, "__enter__", "__exit__"
+                )
+            return NotImplemented
+
+
+class TupleContextManager(tuple, AbstractContextManager):
+    """Base for context managers that are also a tuple."""
+
+    # This is needed to establish a workable MRO.
+
+
+class StdioTuple(TupleContextManager):
+    """Tuple context manager of stdin, stdout and stderr stream-like objects."""
+
+    def __new__(cls, iterable):
+        """Instantiate new tuple from iterable containing three streams."""
+        items = list(iterable)
+        assert len(items) == 3, "iterable must be three items"  # noqa: S101
+
+        return super(StdioTuple, cls).__new__(cls, items)
+
+    @property
+    def stdin(self):
+        """Return capturing stdin stream."""
+        return self[0]
+
+    @property
+    def stdout(self):
+        """Return capturing stdout stream."""
+        return self[1]
+
+    @property
+    def stderr(self):
+        """Return capturing stderr stream."""
+        return self[2]
+
+    def _map_method(self, method: str):
+        """Perform method on all streams."""
+        for item in self:
+            yield item.__dict__[method](item)
+
+    def _map(self, op):
+        """Return generator for performing op on all streams."""
+        if isinstance(op, str):
+            return self._map_method(op)
+
+        return map(op, self)
+
+    def suppress_map(self, ex, op):
+        """Return generator for performing op on all streams, suppressing ex."""
+        for item in self:
+            with suppress(ex):
+                op_inner = item.__dict__[op] if isinstance(op, str) else op
+                yield op_inner(item)
+
+    def _all(self, op):
+        """Perform op on all streams, returning True when all were successful."""
+        return all(self._map(op))
+
+    def suppress_all(self, ex, op):
+        """Perform op on all streams, suppressing ex."""
+        return all(self.suppress_map(ex, op))
+
+    def close(self):
+        """Close all streams."""
+        return self._all("close")
+
+
+class TextIOTuple(StdioTuple):
```

Comment:
this is unused in this patch, and to be honest I dont know how to integrate it yet, except with a factory function which chooses whether to use base StdioTuple or TextIOTuple depending on the environment.  But I suspect that might be the path take, as #78 offers significant features but only if the environment is correct, and often it isnt.